lib/import/colorectal/providers/salisbury/salisbury_handler_colorectal.rb

```diff
@@ -41,98 +44,232 @@ class SalisburyHandlerColorectal < Import::Brca::Core::ProviderHandler
+              dup_genocolorectal.add_status(:negative)
+              genotypes << dup_genocolorectal
             end
             genotypes
           end
 
-          def extract_teststatus(genocolorectal, record)
-            case record.raw_fields['status']
-            when POSITIVE_TEST
-              genocolorectal.add_status(:positive)
-              @logger.debug "POSITIVE status for : #{record.raw_fields['status']}"
-            when /Normal/
-              genocolorectal.add_status(:negative)
-              @logger.debug "POSITIVE status for : #{record.raw_fields['status']}"
-            when /No mutation detected/
-              genocolorectal.add_status(:negative)
-              @logger.debug "POSITIVE status for : #{record.raw_fields['status']}"
-            when /benign/i
-              genocolorectal.add_status(:negative)
-              @logger.debug "POSITIVE status for : #{record.raw_fields['status']}"
-            when FAILED_TEST
-              genocolorectal.add_status(:failed)
-              @logger.debug "FAILED status for : #{record.raw_fields['status']}"
-            else
-              @logger.debug "Cannot determine test status for : #{record.raw_fields['status']}"
+          def process_normal_record(colo_string, genocolorectal, genotypes)
+            colo_string.scan(COLORECTAL_GENES_REGEX).flatten.each do |gene|
+              dup_genocolorectal = genocolorectal.dup_colo
+              dup_genocolorectal.add_gene_colorectal(gene)
+              dup_genocolorectal.add_status(:negative)
+              genotypes << dup_genocolorectal
             end
           end
 
-          def extract_variant(genotype_string, genocolorectal)
-            matches = GENE_LOCATION_REGEX.match(genotype_string)
-            exon_matches = EXON_LOCATION_REGEX.match(genotype_string)
-            if genotype_string.blank?
-              # TODO: what is the desired value to put in here? Negative?
-              genocolorectal.set_negative
-              return
+          def process_failed_record(colo_string, genocolorectal, genotypes)
+            colo_string.scan(COLORECTAL_GENES_REGEX).flatten.each do |gene|
+              dup_genocolorectal = genocolorectal.dup_colo
+              dup_genocolorectal.add_gene_colorectal(gene)
+              dup_genocolorectal.add_status(:failed)
+              genotypes << dup_genocolorectal
             end
-            if matches
-              genocolorectal.add_gene_location(matches[:gene]) if matches[1]
-              genocolorectal.add_protein_impact(matches[:protein]) if matches[2]
-            elsif exon_matches
-              genocolorectal.add_exon_location(exon_matches[1].delete(' '))
-              genocolorectal.add_variant_type(genotype_string)
-            else
-              @logger.warn "Cannot extract gene location from raw test: #{genotype_string}"
+          end
+
+          def process_multiple_tested_genes(genotypes, record, genocolorectal,
```

Comment:
Could you like at refactoring this please, particularly the nested if/else loops?